conveyor/slot.go

```diff
@@ -0,0 +1,238 @@
+	Working
+	Suspending
+)
+
+const slotSize = 10000
+const slotElementDelta = 1000000
+
+// SlotDetails provides information about slot
+type SlotDetails interface {
+	getPulseNumber() core.PulseNumber
+	getNodeId() uint32
+	getPulseData() *core.Pulse
+	getNodeData() interface{}
+}
+
+// HandlersConfiguration contains configuration of handlers for specific pulse state
+// TODO: logic will be provided after pulse change mechanism
+type HandlersConfiguration struct {
+	state SlotState
+}
+
+// TODO: logic will be provided after pulse change mechanism
+func (s *HandlersConfiguration) getMachineConfiguration(smType int) StateMachineType {
+	return nil
+}
+
+// ActivationStatus represents status of work for slot element
+type ActivationStatus int
+
+//go:generate stringer -type=ActivationStatus
+const (
+	EmptyElement = ActivationStatus(iota)
+	ActiveElement
+	NotActiveElement
+)
+
+// ElementList is a list of slotElements with pointers to head and tail
+type ElementList struct {
+	head *slotElement
+	tail *slotElement
+}
+
+// popElement gets element from linked list (and remove it from list)
+func (l *ElementList) popElement() *slotElement {
+	result := l.head
+	if result == nil {
+		return nil
+	}
+	l.head = l.head.listNext
+	return result
+}
+
+// pushElement adds element to linked list
+func (l *ElementList) pushElement(element *slotElement) {
+	if l.head == nil {
+		l.head = element
+		l.tail = element
+	} else {
+		l.tail.listNext = element
+		l.tail = element
+	}
+}
+
+func (l *ElementList) len() int {
+	i := 0
+	for element := l.head; element != nil; element = element.listNext {
+		i++
+	}
+	return i
+}
+
+// Slot holds info about specific pulse and events for it
+type Slot struct {
+	handlersConfiguration HandlersConfiguration
+	inputQueue            queue.IQueue
+	pulseState            PulseState
+	slotState             SlotState
+	pulse                 core.Pulse
+	pulseNumber           core.PulseNumber
+	nodeID                uint32
+	nodeData              interface{}
+	elements              []slotElement
+	elementListMap        map[ActivationStatus]*ElementList
+}
+
+// SlotStateMachine represents state machine of slot itself
+var SlotStateMachine = slotElement{
+	id:               0,
+	state:            0,
+	stateMachineType: 0,
+}
+
+func initElementsBuf() []slotElement {
+	elements := make([]slotElement, slotSize)
+	elements[0] = SlotStateMachine
+	var nextElement *slotElement
+	for i := slotSize - 1; i > 0; i-- {
+		elements[i] = *newSlotElement(EmptyElement)
+		elements[i].id = uint32(i)
+		elements[i].listNext = nextElement
+		nextElement = &elements[i]
+	}
+	return elements
+}
+
+// NewSlot creates new instance of Slot
+func NewSlot(pulseState PulseState, pulseNumber core.PulseNumber) *Slot {
+	slotState := Initializing
+	if pulseState == Antique {
+		slotState = Working
+	}
+
+	elements := initElementsBuf()
+
+	empty := &ElementList{
+		// elements[0] contains SlotStateMachine, so first empty element is elements[1]
```

Comment:
this comment should be in ElementList declaration

Reply:
or it might be better make different field for that element

Reply:
it removes not clear indexes like below ones

Reply:
ElementList don't care about indexes of contained elements.